### PR TITLE
checksafety: better handling of integer shift operators

### LINKED ISSUE #319: Checksafety crash on int shifts
On the following program, the safety checker crashes:

~~~
export
fn main(reg u64 x) -> reg u64 {
  inline int i;
  reg u64 r;
  r = x;
  for i = 0 to (1 << 3) - (5 >> 1) {
    r += x;
  }
  return r;
}
~~~

> Fatal error: exception File "src/safety/safetyInterpreter.ml", line 1863, characters 10-16: Assertion failed

### PR DESCRIPTION
**Pull request by vbgl**
Fixes #319

(The first commit is unrelated, but it does not deserves a PR on its own.)

CI: https://gitlab.com/jasmin-lang/jasmin/-/pipelines/738147944

### CODE REVIEW

**Comment by vbgl** [comment]
Just rebased to fix the merge conflicts in the CHANGELOG.
CI: https://gitlab.com/jasmin-lang/jasmin/-/pipelines/742884901

File: compiler/src/safety/safetyAbsExpr.ml [hunk 1]
```diff
@@ -114,8 +114,10 @@ let op2_to_abs_binop op2 = match op2 with
   | E.Odiv _ -> AB_Arith Texpr1.Div
 
   | E.Olsr _ -> AB_Wop (Wshift Unsigned_right)
-  | E.Olsl _ -> AB_Wop (Wshift Unsigned_left)
-  | E.Oasr _ -> AB_Wop (Wshift Signed_right)
+  | E.Olsl (Op_w _) -> AB_Wop (Wshift Unsigned_left)
+  | E.Olsl Op_int -> AB_Unknown
+  | E.Oasr (Op_w _) -> AB_Wop (Wshift Signed_right)
+  | E.Oasr Op_int -> AB_Unknown
```
Comment (eponier) [inline compiler/src/safety/safetyAbsExpr.ml:120]: Why do we need to distinguish `Op_w` and `Op_int` here? And not for other operations?

**Review by eponier — COMMENTED** [review]
I don't know the safety check, the changes seem reasonable, but I'd like to understand a bit what you did.

File: compiler/src/safety/safetyAbsExpr.ml [hunk 2]
```diff
@@ -324,7 +332,15 @@ module AbsExpr (AbsDom : AbsNumBoolType) = struct
...
     | Papp2 (Odiv Cmp_int, e1, e2) ->
-      obind2 (fun x y -> Some (Z.div x y))
+        obind2 (fun x y -> Some (Z.div x y))
+          (aeval_cst_zint abs e1) (aeval_cst_zint abs e2)
+
+    | Papp2 (Olsl Op_int, e1, e2) ->
+        obind2 zlsl
+          (aeval_cst_zint abs e1) (aeval_cst_zint abs e2)
+
+    | Papp2 (Oasr Op_int, e1, e2) ->
```
[the remark this answers is not in the document]
Comment (vbgl) [reply compiler/src/safety/safetyAbsExpr.ml:342]: There is no `lsr` on `int`.

File: compiler/src/safety/safetyAbsExpr.ml (hunk 1 above)
Comment (vbgl) [reply compiler/src/safety/safetyAbsExpr.ml:120]: I don’t know. Could you provide an example in which not doing this distinction is a problem?

Comment (eponier) [reply compiler/src/safety/safetyAbsExpr.ml:120]: No, I'm trying to understand what you did.

**Review by eponier — APPROVED** [review]
I don't know the safety checker well, so I cannot say that the changes are right, but they are reasonable.

**Pull request merged**